plan/task/clientfilesystem.go

```diff
@@ -0,0 +1,61 @@
+package task
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/gofrs/flock"
+	"github.com/mitchellh/go-homedir"
+	"github.com/rs/zerolog/log"
+	"go.dagger.io/dagger/compiler"
+	"go.dagger.io/dagger/plancontext"
+)
+
+const clientFsTempKey = "client-filesystem-lock"
+
+// clientFsPath extracts the absolute file path from a client filesystem path value.
+func clientFsPath(v *compiler.Value) (string, error) {
+	path, err := v.String()
+	if err != nil {
+		return "", err
+	}
+	expanded, err := homedir.Expand(path)
+	if err != nil {
+		return "", err
+	}
+	return filepath.Abs(expanded)
+}
+
+// clientFsLock acquires a file lock to read/write a client filesystem resource.
+// Every path gets it's own lock. This allows multiple read/write to the same
+// resource, protecting against race conditions.
+func clientFsLock(ctx context.Context, pctx *plancontext.Context, fsPath string) (*flock.Flock, error) {
```

Comment:
A few questions about this:

- Wouldn't it be simpler to just create `{fsPath}.lock` rather than digesting them in a temp dir?
- I get why this is done (multiple writes I assume, also potentially write after read?)
  - I think ideally we'd like to forbid multiple writes to the same path (because the behavior will be random, depending on which write runs first)
  - write after read: I thought you had a workaround for that in CUE?
  - I understand it's not currently possible to implement easily -- maybe worth putting a FIXME somewhere?

Reply:
>  Wouldn't it be simpler to just create `{fsPath}.lock` rather than digesting them in a temp dir?

Not really, because of clean up. 

In my first iteration I put the lock in `cue.mod` but added the buildkit session id to the lock's filename to differenciate between different plan executions. Then I moved that into `~/.config/dagger`.

To clean up, the idea was to remove all lock files starting with the current session's prefix in the filename, after the plan has finished running. I had to put stuff in place in order to do that. Thought about adding CleanUp receiver functions to tasks. But still, if the plan got interrupted before reaching that, you'd be left with locks lying around.

Temporary directories solves both, because they're already in the plan context, they already get removed when the plan is finished, and if execution gets interrupted before that, the OS will remove them on its own.

By doing `{fsPath}.lock`, you'd be creating lock files all over the place, possibly failing to write on a read only permission dir, not isolated from other plan runs so harder to remove afterwards including tracking which ones to `os.Remove()` at the end.

>  I think ideally we'd like to forbid multiple writes to the same path (because the behavior will be random, depending on which write runs first)

The user that originally reported this with `outputs: files` suggested that we should at least fail when detecting this condition. His problem was files getting truncated by concurrent writes. I'm up for that, but locking in gerenal is a **simpler solution**. It also protects against a read while a write is in flight.

And with dynamic tasks, you can't validate this on load, so you'd have to fail after the plan has already executed a few actions. And this is only for files because writing to the same directory has it's use case:

```cue
client: filesystem: {
	lintedGo: write: {
		path: "."
		contents: actions.lint.go.output
	}
	lintedCue: write: {
		path: "."
		contents: actions.lint.cue.output
	}
}
```

So I think locking is just the simpler solution.

> write after read: I thought you had a workaround for that in CUE?

Yes, but this only works inside the same key:

```cue
client: filesystem: "./file": {
    read: contens: string
    write: contents: actions.update.output
}

// or

client: filesystem: kubeconfig: {
    read: {
        path: env.KUBECONFIG
        contens: dagger.#Secret
    }
    write: {
        path: env.KUBECONFIG
        contents: actions.update.output
    }
}
```

But not: 

```
client: filesystem: {
    kuberead: read: {
        path: env.KUBECONFIG
        contens: dagger.#Secret
    }
    kubewrite: write: {
        path: env.KUBECONFIG
        contents: actions.update.output
    }
}
```

> I understand it's not currently possible to implement easily -- maybe worth putting a FIXME somewhere?

You mean failing on **multiple writes** to the same **file**?